File: frontend/types/prediction.ts
export interface ShapValue {
  feature: string;
  display_name: string;
  applicant_value: number;
  shap_contribution: number;
  direction: "increases_risk" | "decreases_risk";
}

export interface PredictionResponse {
  risk_score: number;
  default_probability: number;
  risk_category: "Low" | "Medium" | "High" | "Critical";
  recommendation: "Approve" | "Review" | "Reject";
  confidence: "Low" | "Moderate" | "High";
  shap_values: ShapValue[];
  reasoning: string;
  model_version: string;
}

export interface ModelInfo {
  algorithm: string;
  version: string;
  training_date: string;
  dataset_size: number;
  train_size: number;
  test_size: number;
  features: string[];
  threshold: number;
  scale_pos_weight: number;
  positive_class_rate: number;
  metrics: {
    roc_auc: number;
    precision: number;
    recall: number;
    f1: number;
    accuracy: number;
    confusion_matrix: [[number, number], [number, number]];
    roc_curve: { fpr: number[]; tpr: number[] };
  };
}

export interface GlobalShapFeature {
  feature: string;
  display_name: string;
  importance: number;
}

File: backend/app/schemas/prediction.py
from pydantic import BaseModel, ConfigDict


class ShapValue(BaseModel):
    feature: str
    display_name: str
    applicant_value: float
    shap_contribution: float
    direction: str  # "increases_risk" | "decreases_risk"


class PredictionResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    risk_score: int
    default_probability: float
    risk_category: str
    recommendation: str
    confidence: str
    shap_values: list[ShapValue]
    reasoning: str
    model_version: str
